// phase/client.ts
import Network from 'phase/network';
import User from 'phase/user';
import ElementIdentifiers from 'phase/elementidentifiers';
import PageModel from 'phase/pagemodel';
import Discussion from 'phase/discussion';
import { DiscussionHash, UserHash } from 'phase/hash';
import Message from 'phase/message';

class Client {
    protected id: number;
    protected username: string;
    protected systemName: string;
    protected onlineUsers: Array<User>;
    protected avatar: string;
    protected network: Network;
    protected pageModel: PageModel;
    protected discussions: DiscussionHash;
    protected phaseUsers: UserHash;
    protected currentDiscussion: Discussion | undefined;

    constructor(options: Object) {
        this.discussions = {};
        this.onlineUsers = [];
        this.phaseUsers = {};
        this.network = new Network(this, options);
        this.pageModel = new PageModel(this);
    }

    public setId(id: number): void {
        this.id = id;
    }

    public getId(): number {
        return this.id;
    }

    public setUsername(username: string): void {
        this.username = username;
        this.pageModel.updateUsername(username);
    }

    public getUsername(): string {
        return this.username;
    }

    public setSystemName(systemName: string): void {
        this.systemName = systemName;
    }

    public getSystemName(): string {
        return this.systemName;
    }

    public getOnlineUsers(): Array<User> {
        return this.onlineUsers;
    }

    public getAvatar(): string {
        return this.avatar;
    }

    public getPageModel(): PageModel {
        return this.pageModel;
    }
    
    public getDiscussions(): DiscussionHash {
        return this.discussions;
    }

    public getPhaseUsers(): UserHash {
        return this.phaseUsers;
    }

    /* Updates the users avatar in the chat input area */
    public setAvatar(avatar: string): void {
        this.avatar = avatar;
        this.pageModel.updateAvatar(avatar);
        this.pageModel.refreshUsersOnlineList();
    }

    /* Adds a phase users information to the clients phase users object */
    public addPhaseUser(user: User): void {
        this.getPhaseUsers()[user.getId()] = user;
    }

    /* Appends a discussion to the clients list of discussions and updates page */
    public addDiscussion(discussion: Discussion): void {
        this.discussions[discussion.getId()] = discussion;
        this.pageModel.prependDiscussionsList(discussion);
    }

    /* Switches current discussion
       @param discussionId
            The id of the discussion to switch to
       @param eventElem
            The $(element) that was clicked to cause this event
     */
    public switchDiscussion(discussionId: number, eventElem: any): void {
        let discussion = this.getDiscussions()[discussionId];

        if (!discussion)
            return;

        if (discussion.getMembers().length === 0) {
            this.network.requestDiscussionDetails(discussionId);
        } else {
            this.getPageModel().clearMessages();
            this.getPageModel().displayMessages(discussion);
        }

        if (this.currentDiscussion) {
            this.currentDiscussion.setIsBeingViewed(false);
        }

        discussion.setIsBeingViewed(true);
        this.currentDiscussion = discussion;
        this.getPageModel().switchSelectedDiscussion(eventElem);
        this.getPageModel().setDiscussionRead(discussion);
    }

    public sendChatMessage(text: string): void {
        if (this.currentDiscussion == null) {
            return;
        }

        let chatMessageId = this.network.sendChatMessage(this.currentDiscussion.getId(), text);
        let message = new Message(this.getId(), this.getAvatar(), this.getUsername(), "", text, Math.round(Date.now()/1000), this.getSystemName(), "", "");
        this.currentDiscussion.appendMessage(message);
        this.getPageModel().appendMessage(message);
    }

    public sendInformationRequest(): void {
        let cookie = this.getCookie('_phasesession');
        console.log("cookie: "+cookie);
        if (cookie) {
            this.network.sendInformationRequest(cookie);
        }
    }

    private getCookie(name: string): string | undefined {
        var value = "; " + document.cookie;
        var parts = value.split("; " + name + "=");
        if (parts.length == 2) {
            let element = parts.pop();
            if (element) {
                return element.split(";").shift();
            }
        }

        return "";
    }
}

export default Client;
