// phase/pagemodel.ts
import ElementIdentifiers from 'phase/elementidentifiers';
import Templates from 'phase/templates';
import * as $ from 'jquery';
import Client from 'phase/client';
import Discussion from 'phase/discussion';
import Message from 'phase/message';
import Cookie from 'phase/cookie';

declare var Notification;

class PageModel {
    private client: Client;
    private loginLoadingTimeout: number;
    private windowHasFocus: boolean = true;

    constructor(client: Client) {
        this.client = client;
        this.registerEvents();
    }

    public getWindowHasFocus(): boolean {
        return this.windowHasFocus;
    }

    // Checks to see if a session is available to skip login page
    public checkSession(): void {
        let session = Cookie.getCookie('_phasesession');
        if (session) {
            this.handleLoginSuccess(session);
        }

        let navigate = <any>navigator;
        navigate.serviceWorker.register('sw.js');
    }

    public showNotification(text: string): void {
        console.log("Attempting to show notification for "+text);
        Notification.requestPermission(function(result) {
        let navigate = <any>navigator;
            if (result === 'granted') {
                navigate.serviceWorker.getRegistration().then(function(registration) {
                    registration.showNotification('Phase', {
                        body: text,
                        icon: 'https://t.dark-gaming.com:3001/img/128.png',
                        vibrate: [200, 100, 200, 100, 200, 100, 200],
                        tag: 'vibration-sample'
                    });
                });
            }
        });
    }

    /* Updates the client username in the DOM */
    public updateUsername(username: string): void {
        $(ElementIdentifiers.clientUsername).text(username);
    }

    /* Updates the client avatar in the DOM */
    public updateAvatar(avatar: string): void {
        $(ElementIdentifiers.ChatInput.avatar).html(`<img src="${avatar}" />`);
    }

    /* Clears the existing displayed list and puts in an updated one */
    public refreshUsersOnlineList(): void {
        let sectionList = $(ElementIdentifiers.PeopleSection.main).children(ElementIdentifiers.PeopleSection.list).children('ul');
        sectionList.html('');

        let user;
        let onlineUsers = this.client.getOnlineUsers();
        for (let i = 0; i < onlineUsers.length; i++) {
            user = onlineUsers[i];
            sectionList.append(Templates.PeopleEntry(user));
        }
    }

    /***************************
     * Chat Section / Messages *
     ***************************/

    /* Update the displayed Title */
    public updateDisplayedTitle(name: string): void {
        $(ElementIdentifiers.ChatSection.title).text(name);
    }

    /* Clears the displayed messages */
    public clearMessages(): void {
        $(ElementIdentifiers.ChatSection.list).children().remove();
    }

    /* Displays 40 messages in the chat list from the specified discussion */
    public displayMessages(discussion: Discussion): void {
        let messages = discussion.getMessages();
        let minIndex = messages.length-41 < 0 ? 0 : messages.length-41;
        for (let i = minIndex; i < messages.length; i++) {
            $(ElementIdentifiers.ChatSection.list).append(Templates.ChatMessage(this.client, messages[i]));
        }

        // Scroll the chat to the bottom
        let chatElement = $(ElementIdentifiers.ChatSection.scrollSection);
        chatElement[0].scrollTop = chatElement[0].scrollHeight;
    }

    // Remove the oldest n messages from the discussion message list
    public removeOldestMessages(amount: number): void {
        $(ElementIdentifiers.ChatSection.list).children().slice(0, amount).remove();
    }

    // Add message to the end of the message list
    public appendMessage(message: Message): void {
        let chatElement = $(ElementIdentifiers.ChatSection.scrollSection);
        let maximumScroll = chatElement[0].scrollHeight - chatElement[0].offsetHeight;
        let currentScroll = chatElement[0].scrollTop;
        let scrollIsAtBottom = currentScroll == maximumScroll;

        // Add message to list
        $(ElementIdentifiers.ChatSection.list).append(Templates.ChatMessage(this.client, message));   

        // Scroll the chat to the bottom if they were at the bottom before
        if (scrollIsAtBottom) {
            // Remove messages that go over max of 40 in display
            this.removeOldestMessages($(ElementIdentifiers.ChatSection.list).children().length-40);
            this.scrollToBottom();
        }
    }

    // Scroll the message list to the bottom
    public scrollToBottom(): void {
        let chatElement = $(ElementIdentifiers.ChatSection.scrollSection);
        let maximumScroll = chatElement[0].scrollHeight - chatElement[0].offsetHeight;
        chatElement[0].scrollTop = maximumScroll;
    }
    
    // Creates a live typing element at the bottom of the message list
    public createLiveTyping(message: Message): void {
        $(ElementIdentifiers.ChatSection.liveTyping).append(Templates.LiveTypingMessage(this.client, message));
    }
    
    // Updates the content of an existing live typing element
    public updateLiveTyping(message: Message): void {
        $(`#livetyping-user-${message.getUserId()}`).find('.chat-message-content').text(message.getContentStyled());
    }
    
    // Removes an existing live typing element based on userId
    public removeLiveTyping(userId: number): void {
        $(`#livetyping-user-${userId)}`).remove();
    }

    /*******************
     * Discussion List *
     *******************/

    /* Adds a discussion to the top of the discussions list */
    public prependDiscussionsList(discussion: Discussion): void {
        $(ElementIdentifiers.DiscussionSection.list).prepend(Templates.DiscussionListEntry(discussion));
    }

    /* Updates the selected class to the newly switched discussion */
    public switchSelectedDiscussion(elem: any): void {
        $('.selected').removeClass('selected');
        elem.addClass('selected');
    }

    // Remove unread style to discussion div
    public setDiscussionRead(discussion: Discussion): void {
        $(`#disc${discussion.getId()}`)
            .removeClass('unread');
    }

    // Apply unread style to discussion div
    public setDiscussionUnread(discussion: Discussion): void {
        $(`#disc${discussion.getId()}`)
            .addClass('unread');
    }

    // Update the preview of the discussion div in the discussion list
    public updateDiscussionEntryMessagePreview(discussion: Discussion): void {
        $(`#disc${discussion.getId()}`)
            .children(ElementIdentifiers.DiscussionSection.preview)
            .text(`${discussion.getRecentMessage().getDisplayName().asIs()}: ${discussion.getRecentMessage().getContent().withBbcodeStripped()}`);
    }

    /**********
     * EVENTS *
     **********/

    // Registers the necessary events for interactivity
    protected registerEvents(): void {
        this.registerDiscussionListClick();
        this.registerMessageSend();
        this.registerLogin();
        this.registerWindowFocus();
    } 

    // Registers a click on a discussion list element for switching discussions
    private registerDiscussionListClick(): void {
        let self = this;
        $(ElementIdentifiers.DiscussionSection.list).on('mousedown', ElementIdentifiers.DiscussionSection.discussion, function() {
            self.client.switchDiscussion(+$(this).attr('id').substr(4), $(this));
        });
    }

    // Registers the submit event of the chat message form for sending a message
    private registerMessageSend(): void {
        let self = this;
        $(ElementIdentifiers.ChatInput.textInput).on('keydown', function(event) {
            if (event.keyCode == 13) {
                self.handleMessageSend(this);
            }
        });
    }

    // Tells the client to take the nessary steps to send the message
    private handleMessageSend(context: any): void {
        let chatText = $(context).val();
        this.client.sendChatMessage(chatText);
        $(context).val('');
    }

    // Sends an ajax request using the login details provided when the login form is submitted
    private registerLogin(): void {
        $('#login_form').submit((e) => {
            e.preventDefault();
            
            $.ajax({
                method: "POST",
                url: "https://t.dark-gaming.com:3001/login",
                data: {
                username: $('#username').val(),
                password: $('#password').val(),
                server: $('#servername').find(':selected').text()
                },
                success: (data, textStatus, request) => {
                    this.handleLoginResponse(data, textStatus, request);
                }
            });
            return false;
        });
    }

    // Sets a timeout to remove the loading icon
    private setLoadingTimeout(): void {
        this.loginLoadingTimeout = setTimeout(() => {
            $('.loading').show();
            $('#login_container').addClass('inactive');
        }, 300);
    }

    // Displays a message when there was an error or closes the login form if it was successful
    private handleLoginResponse(data: {state: string, session:string}, textStatus: string, request: any): void {
        switch (data.state) {
            case "success":
                this.handleLoginSuccess(data.session);
                break;
            case "failure":
                $('#login_information').html('<span style="color: Red;">Invalid User/Pass Combination.</span>');
                break;
            case "no-body":
                $('#login_information').html('<span style="color: Red;">This form is invalid. Refresh the page.</span>');
                break;
            case "timeout":
                $('#login_information').html('<span style="color: Red;">The server selected is currently unresponsive.</span>');
                break;
            default:
                $('#login_information').html('<span style="color: Red;">Unknown error.</span>');
                break;
        }

        if (data.state !== 'success') {
            clearTimeout(this.loginLoadingTimeout);
            $('#login_container').removeClass('inactive');
            $('.loading').hide();

            setTimeout(() => {
                $('#login_information').html('');
            }, 4000);
        }
    }

    // Hides the login form and sends off an information request through the websocket
    private handleLoginSuccess(session: string): void {
        $('#phase').css("display", "block");
        $('#login_main').css("display", "none");
        this.client.sendInformationRequest(session);
    }

    // Stores whether the window is in focus
    private registerWindowFocus(): void {
        $(window).focus(() => {
            this.windowHasFocus = true;
        }).blur(() => {
            this.windowHasFocus = false;
        });
    }
}

export default PageModel;
