// phaseclient/client.ts
import Config from "phaseclient/config";
import Cookie from "phaseclient/cookie";
import Discussion from "phaseclient/discussion";
import { BooleanHash, DiscussionHash, TerrariaUserHash, UserHash } from "phaseclient/hash";
import Message from "phaseclient/message";
import Network from "phaseclient/network";
import TerrariaUser from "phaseclient/terrariauser";
import User from "phaseclient/user";
import EditQueue from "phasecore/editqueue";
import DiscussionAlertList from "phasecore/messagetypes/discussionalertlist";
import {Discussion as ReceivedDiscussion, Discussions} from "phasecore/messagetypes/discussions";
import TerrariaUsersOnline from "phasecore/messagetypes/terrariausersonline";
import System from "phasecore/system";
import UserStatus from "phasecore/userstatus";

/**
 * The barebones client. It can be used to connect to the Phase server and has methods for interacting
 * but does not actually do anything useful on its own.
 */
class Client {
    private _uniqueMessageId: number;
    protected id: number;
    protected username: string;
    protected systemName: string;
    protected onlineUsers: User[];
    protected avatar: string;
    protected network: Network;
    protected discussions: DiscussionHash;
    protected phaseUsers: UserHash;
    protected terrariaUsers: TerrariaUserHash;
    protected currentDiscussion: Discussion | undefined;
    protected configuration: Config;
    protected session: string;
    private discussionAlertList: DiscussionAlertList | undefined;
    public switchDiscussionTo: number | null;

    constructor(configuration: Config) {
        this._uniqueMessageId = 0;
        this.id = -1;
        this.discussions = {};
        this.onlineUsers = [];
        this.phaseUsers = {};
        this.terrariaUsers = {};
        this.configuration = configuration;
        this.session = "";
        this.switchDiscussionTo = null;
    }

    public setupNetwork(network: Network): void {
        if (!this.network) {
            this.network = network;
        }
    }

    public getNewMessageId(): number {
        return this._uniqueMessageId++;
    }

    public setId(id: number): void {
        this.id = id;
    }

    public getId(): number {
        return this.id;
    }

    public setUsername(username: string): void {
        this.username = username;
    }

    public getUsername(): string {
        return this.username;
    }

    public setSystemName(systemName: string): void {
        this.systemName = systemName;
    }

    public getSystemName(): string {
        return this.systemName;
    }

    public getOnlineUsers(): User[] {
        return this.onlineUsers;
    }

    public getAvatar(): string {
        return this.avatar;
    }

    public getDiscussions(): DiscussionHash {
        return this.discussions;
    }

    public getCurrentDiscussion(): Discussion | undefined {
        return this.currentDiscussion;
    }

    public getPhaseUsers(): UserHash {
        return this.phaseUsers;
    }

    public getPhaseUser(userId: number): User | undefined {
        return this.getPhaseUsers()[userId.toString()];
    }

    /**
     * Finds a phase user by their username in the list, or null
     * if they do not exist.
     *
     * @param username The username of the Phase User to find
     */
    public getPhaseUserByName(username: string): User | null {
        let foundUser: User | null = null;
        const users = this.getPhaseUsers();
        for (const index in users) {
            if (users[index].getName() === username) {
                foundUser = users[index];
                break;
            }
        }

        return foundUser;
    }

    public getTerrariaUsers(): TerrariaUserHash {
        return this.terrariaUsers;
    }

    public getConfiguration(): Config {
        return this.configuration;
    }

    public getSession(): string {
        return this.session;
    }

    /* Updates the users avatar in the chat input area */
    public setAvatar(avatar: string): void {
        if (avatar.indexOf("/") === 0) {
            avatar = `https://t.dark-gaming.com:3001/${avatar}`;
        }
        this.avatar = avatar;
    }

    /* Adds a phase users information to the clients phase users object */
    public addPhaseUser(user: User): void {
        this.getPhaseUsers()[user.getId()] = user;
    }

    /**
     * A hook for when the discussion list has been received
     */
    protected onDiscussionListReceived(discussions: Discussion[]): void {
        return;
    }

    /* Appends a discussion to the clients list of discussions and updates page */
    public addDiscussion(discussion: Discussion): void {
        this.discussions[discussion.getId()] = discussion;
    }

    /**
     * Switches current discussion
     * @param discussionId The id of the discussion to switch to
     */
    public switchDiscussion(discussionId: number): void {
        const discussion = this.getDiscussions()[discussionId];

        if (!discussion) {
            return;
        }

        if (discussion.getMembers().length === 0) {
            this.network.requestDiscussionDetails(discussionId, this.getConfiguration().useAndDisplayJoinLeaveMessages);
        }

        if (this.currentDiscussion) {
            this.currentDiscussion.setIsBeingViewed(false);
        }

        discussion.setIsBeingViewed(true);
        discussion.setHasUnreadMessages(false);
        this.currentDiscussion = discussion;
    }

    /**
     * Clears the hash of discussions
     */
    public clearDiscussions(): void {
        this.discussions = {};
    }

    /**
     * Sends a queue of edits to the server
     *
     * @param editQueue The queue of edits to send
     */
    public sendEditQueue(editQueue: EditQueue): void {
        if (!this.currentDiscussion) {
            return;
        }

        this.network.sendEditQueue(this.currentDiscussion.getId(), editQueue);
    }

    /**
     * When a chat message is to be displayed it will go through this function.
     * It must be decided whether the message is allowed to be displayed according
     * to the configuration.
     *
     * @param discussion The discussion the message was received for
     * @param message The message itself
     */
    public handleChatMessage(discussion: Discussion, message: Message): void {
        if (!this.getConfiguration().useAndDisplayJoinLeaveMessages
            && message.getUserId() === System.joinLeaveMessageUserId
            && message.getTag().length > 0) {
            return;
        }

        discussion.appendMessage(message);
        discussion.removeLiveTypingMessage(message.getUserId());
        discussion.setRecentMessage(message);

        if (!discussion.getIsBeingViewed()) {
            discussion.setHasUnreadMessages(true);
        }
    }

    /**
     * When the user sends a chat message, it is sent to the server, and a new one put into and is displayed
     * in the discussion list
     *
     * @param text The content of the chat message to send
     */
    public sendChatMessage(text: string): void {
        if (!this.currentDiscussion) {
            return;
        }

        const chatMessageId = this.network.sendChatMessage(this.currentDiscussion.getId(), text);
        const message = new Message(this.getNewMessageId(),
            this.getId(),
            this.getAvatar(),
            this.getUsername(),
            "",
            text,
            Math.round(Date.now() / 1000),
            this.getSystemName(),
            "",
            "");

        // Only process it as a chat message if it is not a command
        if (message.getContent().asIs().indexOf("/") !== 0) {
            this.handleChatMessage(this.currentDiscussion, message);
        }
    }

    /**
     * Use the session to get client information (userId, username etc) necessary
     *
     * @param session The session token
     */
    public sendInformationRequest(session: string): void {
        Cookie.setCookie("_phasesession", session, 3);
        if (session) {
            this.session = session;
            this.network.sendInformationRequest(session);
        }
    }

    /**
     * Request guest access for Phase
     *
     * @param systemName The name of the system to guest
     */
    public sendGuestRequest(systemName: string): void {
        this.network.sendGuestRequest(systemName);
    }

    /**
     * Handles when the discussion messages requested were received or when the local messages were already there and need loading
     *
     * @param discussion The discussion that the messages received are in or the discussion with the local messages already
     */
    public completeDiscussionSwitch(discussion: Discussion): void {
        return;
    }

    /**
     * Handles when a user comes online
     *
     * @param user The user who has come online
     */
    public handleUserOnline(user: User): void {
        if (this.onlineUsers.indexOf(user) === -1) {
            this.onlineUsers.push(user);
        }

        user.setStatus(UserStatus.Online);
    }

    /**
     * Handles when a user goes offline
     *
     * @param user The user who has gone offline
     */
    public handleUserOffline(user: User): void {
        const index = this.onlineUsers.indexOf(user);
        if (index > -1) {
            this.onlineUsers.splice(index, 1);
        }

        user.setStatus(UserStatus.Offline);
        user.setLastOnline(Date.now());

        for (const key in this.discussions) {
            if (typeof this.discussions[key] !== "undefined") {
                this.discussions[key].removeLiveTypingMessage(user.getId());
            }
        }
    }

    /**
     * Handles when the list of users online are received
     *
     * @param userList The users who are online
     */
    public handleUsersOnline(userList: User[]): void {
        this.onlineUsers = [];
        const phaseUsers = this.getPhaseUsers();
        for (const user of userList) {
            if (!phaseUsers[user.getId()]) {
                this.addPhaseUser(user);
            }

            this.onlineUsers.push(user);
        }
    }

    public handleDiscussionList(receivedDiscussions: Discussions): void {
        let discussionInfo: ReceivedDiscussion;
        let message: Message;
        let discussion: Discussion;
        let lastDiscussion: Discussion | null = null;
        const discussions: Discussion[] = [];
        for (let i = receivedDiscussions.length - 1; i >= 0; i--) {
            discussionInfo = receivedDiscussions[i];
            message = new Message(this.getNewMessageId(),
                0,
                "",
                discussionInfo.Username || "",
                discussionInfo.GuestName || "",
                discussionInfo.Message || "[No Content]",
                discussionInfo.timestamp,
                this.getSystemName(),
                "",
                "");
            discussion = new Discussion(discussionInfo.ID, discussionInfo.Name, discussionInfo.Creator, [], message);
            this.addDiscussion(discussion);
            discussions.push(discussion);
            lastDiscussion = discussion;
        }

        this.onDiscussionListReceived(discussions);

        // If the client was reconnected and was previously on a discussion, it is reloaded
        if (this.switchDiscussionTo !== null) {
            this.switchDiscussion(this.switchDiscussionTo);
            this.switchDiscussionTo = null;
        } else {
            if (lastDiscussion) {
                this.switchDiscussion(lastDiscussion.getId());
            }
        }

        const discussionList = this.getDiscussions();
        if (this.discussionAlertList) {
            for (const alert of this.discussionAlertList) {
                if (discussionList[alert.DiscussionID]) {
                    discussionList[alert.DiscussionID].setAlertsEnabled(alert.Alert);
                }
            }
        }
    }

    /**
     * Handles when the client receives a list of which discussions have alerts enabled
     *
     * @param discussionAlertList The list of which discussions have alerts enabled
     */
    public handleDiscussionAlertList(discussionAlertList: DiscussionAlertList): void {
        this.discussionAlertList = discussionAlertList;
    }

    /**
     * Handles when the client becomes a part of a new discussion
     *
     * @param discussion The new discussion
     */
    public handleNewDiscussion(discussion: Discussion): void {
        this.addDiscussion(discussion);
    }

    /**
     * Handles when the client is removed from a discussion
     *
     * @param discussionId The id of the discussion to remove
     */
    public handleRemoveDiscussion(discussionId: number): void {
        delete this.getDiscussions()[discussionId];
    }

    /**
     * Handles when the discussion users list has been received
     *
     * @param discussion The discussion that the users are from
     * @param userList The list of users in the discussion
     * @param modList The moderators of that discussion
     */
     public handleDiscussionUsers(discussion: Discussion, userList: User[], modList: BooleanHash): void {
         discussion.setMembers(userList);
         discussion.setModerators(modList);
     }

    /**
     * Handles when a new live typing message is received
     *
     * @param message The live typing message
     */
    public handleNewLiveTyping(discussion: Discussion, message: Message): void {
        return;
    }

    /**
     * Handles when an update to an existing live typing message is received
     *
     * @param message The live typing message
     */
    public handleUpdateLiveTyping(discussion: Discussion, message: Message): void {
        return;
    }

    /**
     * Handles when a removal of an existing live typing message is received
     *
     * @param userId The user id of the live typing message to remove
     */
    public handleRemoveLiveTyping(discussion: Discussion, userId: number): void {
        return;
    }

    /**
     * Handles when the list of terraria users who are online is received
     *
     * @param users The terraria users who are online
     */
    public handleTerrariaUsers(users: TerrariaUsersOnline): void {
        for (const user of users) {
            this.terrariaUsers[user.name] = new TerrariaUser(user.name, user.tagColour);
        }
    }

    /**
     * Handles when a user joins a terraria server
     *
     * @param username The username of the person who joined a terraria server
     */
    public handleTerrariaJoin(username: string, tagColor: string): void {
        this.terrariaUsers[username] = new TerrariaUser(username, tagColor);
    }

    /**
     * Handles when a user leaves a terraria server
     *
     * @param username The username of the person who left a terraria server
     */
    public handleTerrariaLeave(username: string): void {
        delete this.terrariaUsers[username];

        const user = this.getPhaseUserByName(username);
        if (user !== null) {
            user.setLastInGame(Date.now());
        }
    }

    /**
     * Handles when a user changes their name on a terraria server
     *
     * @param oldUsername The old name of the user
     * @param newUsername The new name of the user
     */
    public handleTerrariaChangeName(oldUsername: string, newUsername: string): void {
        let color = "#000000";
        if (this.terrariaUsers[oldUsername]) {
            color = this.terrariaUsers[oldUsername].getTagColour();
        }

        this.handleTerrariaLeave(oldUsername);
        this.handleTerrariaJoin(newUsername, "#000000");
    }
}

export default Client;
