// phaseclient/deephistory.ts
import Discussion from "phaseclient/discussion";
import Message from "phaseclient/message";

class DeepHistory {
    private _oldestId: number;
    private _newestId: number;
    private _messages: Message[] = [];
    private _discussion: Discussion;
    private _rootMessage: Message;
    private _retrievalInProgress: boolean;
    private _obtainedOldest: boolean = false;

    constructor(discussion: Discussion, rootMessage: Message) {
        this._discussion = discussion;
        this._rootMessage = rootMessage;
    }

    public get oldestId(): number {
        return this._oldestId;
    }

    public get newestId(): number {
        return this._newestId;
    }

    public get messages(): Message[] {
        return this._messages;
    }

    public get discussion(): Discussion {
        return this._discussion;
    }

    public get rootMessage(): Message {
        return this._rootMessage;
    }

    public get retrievalInProgress(): boolean {
        return this._retrievalInProgress;
    }

    public get obtainedOldest(): boolean {
        return this._obtainedOldest;
    }

    public set oldestId(oldestId: number) {
        this._oldestId = oldestId;
    }

    public set newestId(newestId: number) {
        this._newestId = newestId;
    }

    public set retrievalInProgress(retrievalInProgress: boolean) {
        this._retrievalInProgress = retrievalInProgress;
    }

    public set obtainedOldest(obtainedOldest: boolean): void {
        this._obtainedOldest = obtainedOldest;
    }

    public addMessages(messages: Message[]): void {
        this._messages = this._messages.concat(messages);
    }

    public prependMessages(messages: Message[]): void {
        this._messages = messages.concat(this._messages);
    }
}

export default DeepHistory;
