// phaseclient/discussion.ts
import CoreDiscussion from 'phasecore/discussion';
import User from 'phasecore/user';
import { UserHash, MessageHash, BooleanHash } from 'phaseclient/hash';
import Message from 'phaseclient/message';

class Discussion extends CoreDiscussion {
    private recentMessage: Message;
    private ourLiveTypingMessage: string | null;
    private oldestMessageIdReceived: number;
    private isBeingViewed: boolean;
    private hasUnreadMessages: boolean;
    private alertsEnabled: boolean;
    
    constructor(id: number, name: string, creator: number, members: Array<User>, recentMessage: Message) {
        super(id, name, creator, members);
        this.recentMessage = recentMessage;
        this.ourLiveTypingMessage = null;
        this.oldestMessageIdReceived = -1;
        this.isBeingViewed = false;
        this.hasUnreadMessages = false;
        this.alertsEnabled = false;
    }

    public getRecentMessage(): Message {
        return this.recentMessage;
    }

    public getOurLiveTypingMessage(): string | null {
        return this.ourLiveTypingMessage;
    }

    public getIsBeingViewed(): boolean {
        return this.isBeingViewed;
    }

    public getOldestMessageIdReceived(): number {
        return this.oldestMessageIdReceived;
    }

    public getHasUnreadMessages(): boolean {
        return this.hasUnreadMessages;
    }

    public getHasAlertsEnabled(): boolean {
        return this.alertsEnabled;
    }

    public setOurLiveTypingMessage(message: string | null): void {
        this.ourLiveTypingMessage = message;
    }

    public setRecentMessage(recentMessage: Message): void {
        this.recentMessage = recentMessage;
    }

    public appendMessage(message: Message): void {
        super.appendMessage(message);

        // Don't store too many messages
        if (this.messages.length > 200)
            this.messages.shift();
    }

    public setOldestMessageIdReceived(messageId: number): void {
        this.oldestMessageIdReceived = messageId;
    }

    public setIsBeingViewed(isBeingViewed: boolean): void {
        this.isBeingViewed = isBeingViewed;
    }

    public setHasUnreadMessages(hasUnreadMessages: boolean): void {
        this.hasUnreadMessages = hasUnreadMessages;
    }

    public setAlertsEnabled(enabled: boolean): void {
        this.alertsEnabled = enabled;
    }
}

export default Discussion;
