// phaseclient/discussion.ts
import CoreDiscussion from 'phasecore/discussion';
import User from 'phasecore/user';
import { UserHash, MessageHash, BooleanHash } from 'phaseclient/hash';
import Message from 'phaseclient/message';

class Discussion extends CoreDiscussion {
    private recentMessage: Message;
    private ourLiveTypingMessage: string | null;
    private oldestMessageIdReceived: number;
    private isBeingViewed: boolean;
    
    
    constructor(id: number, name: string, creator: number, members: Array<User>, recentMessage: Message) {
        super(id, name, creator, members);
        this.recentMessage = recentMessage;
        this.ourLiveTypingMessage = null;
        this.oldestMessageIdReceived = -1;
        this.isBeingViewed = false;
    }

    public getRecentMessage(): Message {
        return this.recentMessage;
    }

    public getOurLiveTypingMessage(): string | null {
        return this.ourLiveTypingMessage;
    }

    public getIsBeingViewed(): boolean {
        return this.isBeingViewed;
    }

    public getOldestMessageIdReceived(): number {
        return this.oldestMessageIdReceived;
    }

    public setOurLiveTypingMessage(message: string | null): void {
        this.ourLiveTypingMessage = message;
    }

    public setRecentMessage(recentMessage: Message): void {
        this.recentMessage = recentMessage;
    }

    public setOldestMessageIdReceived(messageId: number): void {
        this.oldestMessageIdReceived = messageId;
    }

    public setIsBeingViewed(isBeingViewed: boolean): void {
        this.isBeingViewed = isBeingViewed;
    }
}

export default Discussion;
