// phaseclient/message.ts
import CoreMessage from 'phasecore/message';
import * as Moment from 'moment';
import Utils from 'phasecore/utils';
import SafeString from 'phaseclient/safestring';
import BBCode from 'phaseclient/bbcode';

class Message extends CoreMessage {
    protected isInProgress: boolean;
    private formattedTimestamp: string;
    private contentStyled: string;

    constructor(userId: number, avatar: string, username: string, guestname: string, content: string, timestamp: number, systemName: string, tag: string, tagColour: string) {
        super(userId, avatar, username, guestname, content, timestamp, systemName, tag, tagColour);
        
        this.username = new SafeString(username);
        this.guestname = new SafeString(guestname);
        this.content = new SafeString(content);
        this.isInProgress = false;
    }

    /* Gets a timestamp like "Saturday at 7:16 PM" */
    public getFormattedTimestamp(): string {
        if (!this.formattedTimestamp)  {
            this.formattedTimestamp = Moment(this.getTimestamp()*1000).calendar("", {
                sameDay: "[Today at] LT",
                lastWeek: "dddd [at] LT",
                sameElse: "LT, dddd Do MMM, YYYY"
            });
        }

        return this.formattedTimestamp;
    }

    public setContent(content: string): void {
        super.setContent(content);
        this.contentStyled = "";
    }

    /* Get BBCode formatted content */
    public getContentStyled(): string {
        if (!this.contentStyled) {
            this.contentStyled = BBCode.convertText(this.getContent().getContentWithHtmlEscaped());
        }

        return this.contentStyled;
    }

    public getIsInProgress(): boolean {
        return this.isInProgress;
    }
}

export default Message;
