// phaseclient/search.ts
import Discussion from "phaseclient/discussion";
import Message from "phaseclient/message";

class Search {
    private readonly _terms: string;
    private readonly _discussion: Discussion;
    private _oldestId: number | null | undefined;
    private readonly _username: string | undefined;
    private readonly _systemName: string | undefined;
    private _messages: Message[] = [];
    private _inProgress: boolean = true;

    constructor(terms: string, discussion: Discussion, oldestId?: number, username?: string, systemName?: string) {
        this._terms = terms;
        this._discussion = discussion;
        this._oldestId = oldestId;
        this._username = username;
        this._systemName = systemName;
    }

    public get terms(): string {
        return this._terms;
    }

    public get discussion(): Discussion {
        return this._discussion;
    }

    public get messages(): Message[] {
        return this._messages;
    }

    public get oldestId(): number | null | undefined {
        return this._oldestId;
    }

    public get username(): string | undefined {
        return this._username;
    }

    public get systemName(): string | undefined {
        return this._systemName;
    }

    public get inProgress(): boolean {
        return this._inProgress;
    }

    public set oldestId(oldestId: number | null | undefined) {
        this._oldestId = oldestId;
    }

    public set inProgress(inProgress: boolean) {
        this._inProgress = inProgress;
    }

    public clearMessages(): void {
        this._messages = [];
    }

    public addMessages(messages: Message[]): void {
        this._messages = this._messages.concat(messages);
    }

    public prependMessages(messages: Message[]): void {
        this._messages = messages.concat(this._messages);
    }
}

export default Search;
