// phaseclient/user.ts
import CoreUser from 'phasecore/user';
import Utils from 'phasecore/utils';
import UserStatus from 'phasecore/userstatus';

class User extends CoreUser {
    private formatName: string;
    private formatNameNoBbcode: string;
    private lowerCaseName: string;
    private lastOnline: number;
    private lastInGame: number;
    
    constructor(id: number, name: string, systemName: string, avatar: string, clientSystemName: string, status: UserStatus) {
        super(id, name, systemName, avatar, status);

        // Users from other systems will have a suffix to show for it
        if (systemName !== clientSystemName) {
            this.formatName = `${Utils.htmlspecialchars(name)}@<i>${systemName}</i>`;
            this.formatNameNoBbcode = `${Utils.htmlspecialchars(name)}@${systemName}`;
        } else {
            this.formatName = Utils.htmlspecialchars(name);
            this.formatNameNoBbcode = this.formatName;
        }

        this.lowerCaseName = name.toLowerCase();
    }
    
    /**
     * Gets the username combined with an appropriate system name for this user
     */
    public getFormatName(): string {
        return this.formatName;
    }

    /**
     * Gets the username combined with an appropriate system name for this user
     */
    public getFormatNameWithoutBBCode(): string {
        return this.formatName;
    }

    /**
     * Gets the username in lowercase
     */
    public getLowerCaseName(): string {
        return this.lowerCaseName;
    }

    /**
     * Gets the last time this user was online on Phase
     * 
     * @return The unix timestamp of the last time this user was on Phase
     */
    public getLastOnline(): number {
        return this.lastOnline;
    }

    /**
     * Gets the last time this user was online in-game
     * 
     * @return The unix timestamp of the last time this user was in-game
     */
    public getLastInGame(): number {
        return this.lastOnline;
    }

    /**
     * Updates the last time this user was on Phase
     * 
     * @param lastOnline The last time this user was online on Phase
     */
    public setLastOnline(lastOnline: number): void {
        this.lastOnline = lastOnline;
    }

    /**
     * Updates the last time this user was in-game
     * 
     * @param lastInGame The last time this user was online in-game
     */
    public setLastInGame(lastInGame: number): void {
        this.lastInGame = lastInGame;
    }
}

export default User;
