// phase/discussion.ts
import User from 'phase/user';
import { UserHash, MessageHash, BooleanHash } from 'phase/hash';
import Message from 'phase/message';

class Discussion {
    private id: number;
    private name: string;
    private creator: number;
    private members: Array<User>;
    private moderators: BooleanHash;
    private recentMessage: Message;
    private messages: Array<Message>;
    private liveTypingMessages: MessageHash;
    private oldestMessageIdReceived: number;
    private isBeingViewed: boolean;

    constructor(id: number, name: string, creator: number, members: Array<User>, recentMessage: Message) {
        this.id = id;
        this.name = name;
        this.creator = creator;
        this.members = members;
        this.moderators = {};
        this.recentMessage = recentMessage;
        this.messages = [];
        this.liveTypingMessages = {};
        this.oldestMessageIdReceived = -1;
        this.isBeingViewed = false;
    }

    public getId(): number {
        return this.id;
    }

    public getName(): string {
        return this.name;
    }

    public getCreator(): number {
        return this.creator;
    }

    public getMembers(): Array<User> {
        return this.members;
    }

    public getModerators(): BooleanHash {
        return this.moderators;
    }

    public getRecentMessage(): Message {
        return this.recentMessage;
    }

    public getMessages(): Array<Message> {
        return this.messages;
    }

    public getLiveTypingMessages(): MessageHash {
        return this.liveTypingMessages;
    }

    public getLiveTypingMessage(userId: number): Message {
        return this.getLiveTypingMessages()[userId.toString()];
    }

    public getOldestMessageIdReceived(): number {
        return this.oldestMessageIdReceived;
    }

    public getIsBeingViewed(): boolean {
        return this.isBeingViewed;
    }

    public setMembers(members: Array<User>): void {
        this.members = members;
    }

    public setModerators(moderators: BooleanHash): void {
        this.moderators = moderators;
    }

    public setRecentMessage(recentMessage: Message): void {
        this.recentMessage = recentMessage;
    }

    public setMessages(messages: Array<Message>): void {
        this.messages = messages;
    }

    public prependMessage(message: Message): void {
        this.messages.unshift(message);
    }

    public appendMessage(message: Message): void {
        this.messages.push(message);
    }

    public setLiveTypingMessages(userId: number, message: Message): void {
        this.getLiveTypingMessages()[`${userId}`] = message;
    }

    public updateLiveTypingMessageContent(userId: number, content: string): void {
        this.getLiveTypingMessages()[`${userId}`].setContent(content);
    }
    
    public removeLiveTypingMessage(userId: number): void {
        delete this.getLiveTypingMessages()[`${userId}`];
    }

    public setOldestMessageIdReceived(messageId: number): void {
        this.oldestMessageIdReceived = messageId;
    }

    public setIsBeingViewed(isBeingViewed: boolean): void {
        this.isBeingViewed = isBeingViewed;
    }
}

export default Discussion;
