// phase/user.ts
import Utils from 'phase/utils';

class User {
    private id: number;
    private name: string;
    private formatName: string;
    private systemName: string;
    private avatar: string;

    constructor(id: number, name: string, systemName: string, avatar: string, clientSystemName: string) {
        this.id = id;
        this.name = name;
        this.systemName = systemName;
        this.avatar = avatar;

        // Users from other systems will have a suffix to show for it
        if (systemName !== clientSystemName) {
            this.formatName = `${Utils.htmlspecialchars(name)}@<i>${systemName}</i>`;
        } else {
            this.formatName = Utils.htmlspecialchars(name)
        }
    }

    public getId(): number {
        return this.id;
    }

    public getName(): string {
        return this.name;
    }

    public getFormatName(): string {
        return this.formatName;
    }

    public getAvatar(): string {
        return this.avatar;
    }

    public setAvatar(avatar: string): void {
        this.avatar = avatar;
    }
}

export default User;
