// phaseweb/templates/chatmessage.ts
import gibb from "asdfjkl";
import PhaseClient from "phaseclient/client";
import Discussion from "phaseclient/discussion";
import Message from "phaseclient/message";
import ElementIdentifiers from "phaseweb/elementidentifiers";
import Hash from "phaseweb/hash";
import Client from "phaseweb/templates/client";

/**
 * Gets whether the system name is to be used
 *
 * @param client The client
 * @param message The message being displayed
 */
function shouldUseSystemName(client: Client, message: Message) {
    const useSystemName = client.configuration.displayOtherSystemsNames
            && message.systemName !== client.systemName
            && message.systemName !== null;

    return useSystemName;
}
/**
 * Gets the system name in HTML form
 *
 * @param client The client
 * @param message The message being displayed
 */
function getSystemNameTag(client: Client, message: Message): string {
    return shouldUseSystemName(client, message) ? `<span class="systemSuffix">@${message.systemName}</span>` : ``;
}

/**
 * Gets the message tag in HTML form
 *
 * @param message The message being displayed
 */
function getMessageTag(message: Message): string {
    return getUseTag(message) ? `<span style="margin-left: 5px; color: ${message.tagColour}">${message.tag}</span>` : ``;
}

/**
 * Gets whether the message tag is to be displayed
 *
 * @param message The message being displayed
 */
function getUseTag(message: Message): boolean {
    return message.tag.length > 0;
}

/**
 * Gets the online staus of a terraria user's message in HTML form
 *
 * @param client The client
 * @param message The message being displayed
 */
function getTerrariaStatus(client: Client, message: Message): string {
    const displayName = message.getDisplayName().getContentWithHtmlEscaped();
    const isInTerraria = client.terrariaUsers[displayName];
    const terrariaStatusClass = isInTerraria
        ? ElementIdentifiers.Message.terrariaOnline.substr(1)
        : ElementIdentifiers.Message.terrariaOffline.substr(1);
    const statusServer = isInTerraria
        ? ` style="background: ${client.terrariaUsers[displayName].tagColour}"`
        : "";
    let terrariaStatus = "";
    if (getUseTag(message) && !shouldUseSystemName(client, message)) {
        terrariaStatus = `<div class="${terrariaStatusClass} ${ElementIdentifiers.Message.terrariaStatus.substr(1)} `
          + `chat-terraria-${Hash(displayName)}"${statusServer}></div>`;
    }

    return terrariaStatus;
}

function getIpAddressElement(message: Message): string {
    if (message.ip.length > 0) {
        return `<span class="chat-message-ip">${message.ip}</span>`;
    }

    return "";
}

/**
 * Gets whether to use the alert styling on the message
 *
 * @param client The client
 * @param chatMessage The chat message text
 */
function isAlertMessage(client: Client, chatMessage: string): boolean {
    return chatMessage.indexOf("@" + client.username) > -1 || chatMessage.indexOf("@everyone") > -1 || chatMessage.indexOf("@here") > -1;
}

function isSpamMessage(message: Message) {
    return message.content.asIs().length > 10 ? gibb(message.content.asIs()) : false;
}

/**
 * Gets the alert symbol as HTML
 *
 * @param client The client
 * @param chatMessage  The chat message text
 */
function getAlertSymbolTag(client: Client, chatMessage: string): string {
    return isAlertMessage(client, chatMessage) ? `<span class="chat-alert"><i class="fa fa-exclamation-circle" aria- hidden="true" ></i></span>` : ``;
}

/**
 * Gets the spam alert symbol as HTML
 *
 * @param chatMessage  The chat message
 */
function getSpamSymbolTag(message: Message): string {
    return isSpamMessage(message) ? `<span class="chat-spam-alert">Spam?</span>` : ``;
}

/**
 * Adds styling to the text that caused an alert
 *
 * @param chatMessage The chat message text
 */
function addAlertStyling(client: Client, chatMessage: string): string {
    return chatMessage.replace("@" + client.username, `@<span class="alert-name">${client.username}</span>`);
}

export default function ChatMessage(client: Client, discussion: Discussion, message: Message) {
    const accountName = message.guestName.asIs().length > 0 && message.username.asIs().length > 0 ? message.username : "";
    const displayName = message.getDisplayName().getContentWithHtmlEscaped();
    // Remove chat tags from name if they exist to "clean it up"
    const strippedName = displayName.replace(/\[(?:.*?):(.*?)]/g, "$1");
    const nameWasStripped = strippedName !== displayName;
    const visualStrippedIcon = nameWasStripped
        ? `<span class="chat-name-asterisk" title="${displayName}">*</span>`
        : "";

    const systemName = message.systemName;
    const tag = message.tag;
    const tagColour = message.tagColour;
    const chatMessage = addAlertStyling(client, message.getContentStyled(discussion));

    // Flags
    const isCreator = message.userId === discussion.creator;
    const isMod = discussion.moderators[message.userId];
    const isGuest = message.username.asIs().length === 0 && message.guestName.asIs().length > 0;

    const guestClass = isGuest ? ` ${ElementIdentifiers.Message.guest.substr(1)}` : "";
    const creatorClass = isCreator ? ` ${ElementIdentifiers.Message.creator.substr(1)}` : "";
    const modClass = isMod ? ` ${ElementIdentifiers.Message.moderator.substr(1)}` : "";
    const creatorOrModClass = isCreator ? creatorClass : modClass;

    let templatedMessage = "";

    if (client.configuration.shrinkJoinLeaveMessages && message.userId === -2) {
        templatedMessage = `
        <li id="${message.id}" class="local${message.localMessageId}">
            <div class="${ElementIdentifiers.Message.avatar.substr(1)}">
                <div class="${ElementIdentifiers.Message.info.substr(1)}">`
                    + `${getMessageTag(message)}`
                + `</div>
            </div>
            <div class="chat-message-right">
                <div class="chat-message-content">${chatMessage}</div>
            </div>
            <div style="clear: both; content: ' '; height: 0"></div>
        </li>`;
    } else {
        templatedMessage = `
        <li id="${message.id}" class="local${message.localMessageId}">
            <div class="${ElementIdentifiers.Message.avatar.substr(1)}">
                <img class="avatar_${message.userId}"
                     src="${message.avatar}"
                     onError="this.onerror=null;this.src='https://t.dark-gaming.com:3001/img/question-mark.jpg';">
                ${getTerrariaStatus(client, message)}
            </div>
            <div class="chat-message-right">
                <div class="${ElementIdentifiers.Message.info.substr(1)}">`
                    + `<span class="chat-name${creatorOrModClass}${guestClass}"
                             title="${accountName}">${strippedName}${visualStrippedIcon}</span>`
                    + `${(getSystemNameTag(client, message))}`
                    + `<span class="chat-timestamp">${message.getFormattedTimestamp()}</span>`
                    + `${getMessageTag(message)}`
                    + `${getAlertSymbolTag(client, chatMessage)}`
                    + `${getSpamSymbolTag(message)}`
                    + `${getIpAddressElement(message)}`
                + `</div>
                <div class="chat-message-content"><div class="chat-line">${chatMessage}</div></div>
            </div>
            <div style="clear: both; content: ' '; height: 0"></div>
        </li>`;
    }

    return templatedMessage;
}
