// phaseweb/pagemodel/messagesection/contextmenu.ts
import $ from "jquery";
import Client from "phaseclient/client";
import ElementIdentifiers from "phaseweb/elementidentifiers";
import Translator from "phaseweb/translator";
import WebClient from "phaseweb/client";

class ContextMenu {
    public translator: Translator = new Translator();
    private _contextMenuMessageId: string;
    private _client: Client;

    constructor(client: Client) {
        this._client = client;
    }

    public registerEvents(): void {
        this.registerPageClick();
        this.registerMessageContextMenu();
        this.registerTranslateClick();
        this.registerCopyMessageLinkClick();
        this.registerKickClick();
        this.registerBanClick();
    }

    /**
     * Hides the context menu
     */
    public hide(): void {
        $(ElementIdentifiers.ContextMenu.menu).offset({
            left: -500,
            top: -500
        });
    }

    /**
     * Clears the context menu when the page is clicked
     */
    private registerPageClick(): void {
        $("body").on("click", () => {
            this.hide();
        });
    }

    /**
     * Binds the context menu appearing to right clicking a message entry
     */
    private registerMessageContextMenu(): void {
        const self = this;

        $(ElementIdentifiers.ChatSection.list).on("contextmenu", "li", function(e) {
            const selection = window.getSelection().toString();
            if (!$(this).hasClass("in-progress") && selection.length === 0 && !$(e.target).parent().hasClass("chat-message-info")
                && !$(e.target).is("a") && !$(e.target).is("img")) {
                e.preventDefault();

                // Make sure it doesn't go out of visisble area (under bottom)
                if (window.outerHeight - e.pageY < 265) {
                    e.pageY = window.outerHeight - 265;
                }

                $(ElementIdentifiers.ContextMenu.menu)
                    .offset({
                        left: e.pageX,
                        top: e.pageY
                    });

                self._contextMenuMessageId = $(this).attr("id");
            }
        });
    }

    /**
     * Translates text when the context menu item is clicked
     */
    private registerTranslateClick(): void {
        $(ElementIdentifiers.ContextMenu.translate).on("click", () => {
            const element = $(`#${this._contextMenuMessageId}`).find(ElementIdentifiers.Message.content);
            const lines = element.children().toArray();

            const messageId = this._contextMenuMessageId;
            for (const line of lines) {
                this.translator.translate($(line).text(), "en")
                    .then((translatedMessage: string) => {
                        $(line).text(translatedMessage);
                    })
                    .catch((e: Error) => {
                        console.log(e);
                    });
            }
        });
    }

    private registerCopyMessageLinkClick(): void {
        $(ElementIdentifiers.ContextMenu.messageLink).on("click", () => {
            const elem = $(`<input type="text" value="https://phase.dark-gaming.com/#messageref_${this._contextMenuMessageId}" />`);
            $("body").append(elem);
            (elem[0] as HTMLInputElement).select();
            document.execCommand("copy");
            elem.remove();
        });
    }

    private registerKickClick(): void {
        $(ElementIdentifiers.ContextMenu.kick).on("click", () => {
            if (this._client.currentDiscussion) {
                const message = this._client.currentDiscussion.messages.find(m => m.id === this._contextMenuMessageId);
                if (message) {
                    this._client.currentDiscussion.ourLiveTypingMessage = 
                        `/kick -u "${message.getDisplayName().asIs()}" -s ${message.tag.toLowerCase()} -r ""`;
                    (this._client as WebClient).pageModel.messageSection.updateOurLiveTyping(this._client.currentDiscussion);
                    (this._client as WebClient).pageModel.messageSection.focusChatInput(this._client.currentDiscussion.ourLiveTypingMessage.length - 1);
                }
            }
        });
    }

    private registerBanClick(): void {
        $(ElementIdentifiers.ContextMenu.ban).on("click", () => {
            if (this._client.currentDiscussion) {
                const message = this._client.currentDiscussion.messages.find(m => m.id === this._contextMenuMessageId);
                if (message) {
                    this._client.currentDiscussion.ourLiveTypingMessage = 
                        `/ban -u "${message.getDisplayName().asIs()}" -s build -ip "${message.ip}" -r ""`;
                    (this._client as WebClient).pageModel.messageSection.updateOurLiveTyping(this._client.currentDiscussion);
                    (this._client as WebClient).pageModel.messageSection.focusChatInput(this._client.currentDiscussion.ourLiveTypingMessage.length - 1);
                }
            }
        });
    }
}

export default ContextMenu;
