// phaseweb/pagemodel/index.ts
import * as $ from "jquery";
import Cookie from "phaseclient/cookie";
import Discussion from "phaseclient/discussion";
import Message from "phaseclient/message";
import User from "phaseclient/user";
import Client from "phaseweb/client";
import ElementIdentifiers from "phaseweb/elementidentifiers";
import Hash from "phaseweb/hash";
import DiscussionSection from "phaseweb/pagemodel/discussionsection";
import MessageSection from "phaseweb/pagemodel/messagesection";
import PeopleSection from "phaseweb/pagemodel/peoplesection";

declare var Notification;

class PageModel {
    private _client: Client;
    private _loginLoadingTimeout: any;
    private _windowHasFocus: boolean;
    private _lastNotification: Notification | null;
    private _notificationSound: HTMLAudioElement;
    public peopleSection: PeopleSection;
    public discussionSection: DiscussionSection;
    public messageSection: MessageSection;

    constructor(client: Client) {
        this._client = client;
        this.peopleSection = new PeopleSection(client);
        this.messageSection = new MessageSection(client);
        this.discussionSection = new DiscussionSection(client);
        this._windowHasFocus = true;
        this._lastNotification = null;
        this.registerEvents();
        this.setupScrollers();
        this._notificationSound = new Audio("https://t.dark-gaming.com:3001/sounds/sonar.wav");

        Notification.requestPermission((result) => 0);
    }

    public getWindowHasFocus(): boolean {
        return this._windowHasFocus;
    }

    /**
     * Checks to see if a session is available to skip login page
     */
    public checkSession(): void {
        const session = Cookie.getCookie("_phasesession");
        if (session) {
            this.handleLoginSuccess(session);
        } else {
            $(ElementIdentifiers.Login.loading).hide();
            $(ElementIdentifiers.Login.container).removeClass("inactive");
        }

        const navigate = navigator as any;
        navigate.serviceWorker.register("sw.js");
    }

    /**
     * Displays a notification to the user
     *
     * @param text The text body of the notification
     */
    public showNotification(discussion: Discussion, text: string): void {
        if (!Notification) {
            alert("Notifications are supported in modern versions of Chrome, Firefox, Opera and IE.");
            return;
        }

        if (Notification.permission !== "granted") {
            Notification.requestPermission();
        }

        if (this._lastNotification !== null) {
            this._lastNotification.close();
        }

        const notification = new Notification(discussion.getName(), {
            icon: "https://t.dark-gaming.com:3001/img/128.png",
            body: text,
            id: "darkgamingphase",
            silent: true
        });

        this._lastNotification = notification;
        notification.onclick = () => {
            window.focus();
            this._client.switchDiscussion(discussion.getId());
            this._lastNotification = null;
        };

        this.playNotificationSound();
    }

    /**
     * Plays the notification sound
     */
    private playNotificationSound(): void {
        this._notificationSound.play();
    }

    /**
     * Updates the client username in the DOM
     *
     * @param username The username of the current user
     */
    public updateUsername(username: string): void {
        $(ElementIdentifiers.clientUsername).text(username);
        this.messageSection.updateOurDisplayedName(username);
    }

    /**
     * Updates the client avatar in the DOM
     *
     * @param avatar The avatar of the current user
     */
    public updateAvatar(avatar: string): void {
        $(ElementIdentifiers.ChatInput.avatar).html(`<img src="${avatar}" />`);
    }

    public updateUserAvatar(user: User): void {
        $(`.avatar_${user.getId()}`).attr("src", user.getAvatar());
        $(`.avatar_hash${Hash(user.getFormatNameWithoutBBCode())}`).attr("src", user.getAvatar());
    }

    /**********
     * EVENTS *
     **********/

    /**
     * Registers the necessary events for interactivity
     */
    protected registerEvents(): void {
        this.messageSection.registerEvents();
        this.discussionSection.registerDiscussionListClick();
        this.peopleSection.registerEvents();
        this.registerLogin();
        this.registerWindowFocus();
    }

    /**
     * Sends an ajax request using the login details provided when the login form is submitted
     */
    private registerLogin(): void {
        $("#login_form").submit((e) => {
            e.preventDefault();
            this.setLoadingTimeout();

            $.ajax({
                method: "POST",
                url: "https://t.dark-gaming.com:3001/login",
                data: {
                username: $("#username").val(),
                password: $("#password").val(),
                server: $("#servername").find(":selected").text()
                },
                success: (data, textStatus, request) => {
                    this.handleLoginResponse(data, textStatus, request);
                }
            });
            return false;
        });
    }

    /**
     * Sets a timeout to remove the loading icon
     */
    private setLoadingTimeout(): void {
        this._loginLoadingTimeout = setTimeout(() => {
            $(ElementIdentifiers.Login.loading).show();
            $(ElementIdentifiers.Login.container).addClass("inactive");
        }, 300);
    }

    /**
     * Displays a message when there was an error or closes the login form if it was successful
     *
     * @param data The response state and session token
     * @param textStatus
     * @param request
     */
    private handleLoginResponse(data: {state: string, session: string}, textStatus: string, request: any): void {
        switch (data.state) {
            case "success":
                this.handleLoginSuccess(data.session);
                break;
            case "failure":
                $(ElementIdentifiers.Login.information).html('<span style="color: Red;">Invalid User/Pass Combination.</span>');
                break;
            case "no-body":
                $(ElementIdentifiers.Login.information).html('<span style="color: Red;">This form is invalid. Refresh the page.</span>');
                break;
            case "timeout":
                $(ElementIdentifiers.Login.information).html('<span style="color: Red;">The server selected is currently unresponsive.</span>');
                break;
            default:
                $(ElementIdentifiers.Login.information).html('<span style="color: Red;">Unknown error.</span>');
                break;
        }

        if (data.state !== "success") {
            clearTimeout(this._loginLoadingTimeout);
            $(ElementIdentifiers.Login.container).removeClass("inactive");
            $(ElementIdentifiers.Login.loading).hide();

            setTimeout(() => {
                $(ElementIdentifiers.Login.information).empty();
            }, 4000);
        }
    }

    /**
     * Hides the login form and sends off an information request through the network
     *
     * @param session The session token received from the login request
     */
    private handleLoginSuccess(session: string): void {
        $("#phase").css("display", "block");
        $(ElementIdentifiers.Login.main).css("display", "none");
        this._client.sendInformationRequest(session);
    }

    /**
     * Stores whether the window is in focus
     */
    private registerWindowFocus(): void {
        $(window).focus(() => {
            this._windowHasFocus = true;
        }).blur(() => {
            this._windowHasFocus = false;
        });
    }

    /*******************
    * Custom Scrollers *
    ********************/
    private setupScrollers() {
        this.discussionSection.setupScroller();
        this.peopleSection.setupScrollers();
    }
}

export default PageModel;
